xborders

```diff
@@ -199,85 +177,120 @@ class Highlight(Gtk.Window):
-                    h -= BORDER_WIDTH;
-                elif BORDER_MODE == 1:
-                    x -= BORDER_WIDTH / 2;
-                    y -= BORDER_WIDTH / 2;
-                    w += BORDER_WIDTH;
-                    h += BORDER_WIDTH;
-                draw.draw_rectangle(
-                    ctx,
-                    x,
-                    y,
-                    w,
-                    h,
-                    BORDER_RADIUS,
-                    BORDER_WIDTH,
-                    [0, 0, 0, 0],
-                    [BORDER_R, BORDER_G, BORDER_B, BORDER_A],
-                )
+            self.move(1e6, 1e6)
+            subprocess.Popen(["notify-send", "xborder", "xborders requires a compositor. Resuming once a compositor is running."])
+
+    # This event will trigger every active window change, it will queue a
+    # border to be drawn and then do nothing.
+    # See:
+    #    Signals available for the Wnck.Screen class: https://lazka.github.io/pgi-docs/Wnck-3.0/classes/Screen.html#signals
+    #    Signals available for the Wnck.Window class: https://lazka.github.io/pgi-docs/Wnck-3.0/classes/Window.html#signals
+    def _active_window_changed_event(self, screen, previous_active_window):
+        active_window = self.wnck_screen.get_active_window()
+
+        self.border_path = [0, 0, 0, 0]
+        if active_window is not None:
+            # Find if the window has a 'geometry-changed' event connected.
+            signal_id = GObject.signal_lookup('geometry-changed', active_window)
+            has_event_connected = GObject.signal_has_handler_pending(active_window, signal_id, 0, False)
+
+            # if it doesn't have one.
+            if not has_event_connected:
+                # Connect it.
+                # Has to be done this way in order to not connect an event
+                # every time the active window changes, thus, drawing unnecesary frames.
+                active_window.connect('geometry-changed', self._geometry_changed_event)
+
+            if active_window.get_state() != Wnck.WindowState.FULLSCREEN:
+                self._calc_border_geometry(active_window)
+        self.queue_draw()
 
-        ctx.restore()
+    # This is weird, "window_changed" is not necessarily the active window,
+    # it is the window which receives the signal of resizing and is not necessarily
+    # the active window, this means the border will get drawn on other windows.
+    def _geometry_changed_event(self, window_changed):
+        active_window = self.wnck_screen.get_active_window()
+        if active_window is None or active_window.get_state() == Wnck.WindowState.FULLSCREEN:
+            self.border_path = [0, 0, 0, 0]
+        else:
+            self._calc_border_geometry(active_window)
         self.queue_draw()
 
+    def _calc_border_geometry(self, window):
+        x, y, w, h = window.get_geometry()
 
-def main():
-    get_args()
+        # Center
+        self.border_path = [x, y, w, h]
```

Comment:
Can we not just modify `x, y, w, h` and set this at the end of the function? No need to do `self.border_path[n]` each time

Reply:
I no longer use this variable, see below.